frame/contracts/src/lib.rs

```diff
@@ -1132,111 +1273,47 @@ impl<T: Config> Pallet<T> {
 		self::wasm::reinstrument(module, schedule).map(|_| ())
 	}
 
-	/// Internal function that does the actual call.
+	/// Single entry point to contract execution.
+	/// Downstream execution flow is branched by implementations of [`Invokable`] trait:
 	///
-	/// Called by dispatchables and public functions.
-	fn internal_call(
-		origin: T::AccountId,
-		dest: T::AccountId,
-		value: BalanceOf<T>,
-		gas_limit: Weight,
-		storage_deposit_limit: Option<BalanceOf<T>>,
-		data: Vec<u8>,
+	/// - [`InstantiateInput::run`] runs contract instantiation,
+	/// - [`CallInput::run`] runs contract call.
+	///
+	/// We enforce a re-entrancy guard here by initializing and checking a boolean flag through a
+	/// global reference.
+	fn internal_run<I: Invokable<T>>(
+		input: I,
 		debug_message: Option<&mut DebugBufferVec<T>>,
-		determinism: Determinism,
-	) -> InternalCallOutput<T> {
-		let mut gas_meter = GasMeter::new(gas_limit);
-		let mut storage_meter = match StorageMeter::new(&origin, storage_deposit_limit, value) {
-			Ok(meter) => meter,
-			Err(err) =>
-				return InternalCallOutput {
-					result: Err(err.into()),
-					gas_meter,
-					storage_deposit: Default::default(),
-				},
-		};
-		let schedule = T::Schedule::get();
-		let result = ExecStack::<T, PrefabWasmModule<T>>::run_call(
-			origin.clone(),
-			dest,
-			&mut gas_meter,
-			&mut storage_meter,
-			&schedule,
-			value,
-			data,
-			debug_message,
-			determinism,
-		);
-		InternalCallOutput {
-			result,
-			gas_meter,
-			storage_deposit: storage_meter.into_deposit(&origin),
-		}
-	}
+	) -> I::Output {
+		let gas_meter = GasMeter::new(input.gas_limit());
+		reentrancy_guard::using_once(&mut false, || {
+			let res = reentrancy_guard::with(|f| {
+				// Check re-entrancy guard
+				if *f {
+					return Err(())
+				}
+				// Set re-entracy guard
+				*f = true;
```

Comment:
Those comments do not add any information IMHO. So either just remove or improve. Focusing on the **why** rather than the **what** usually triggers the right thoughts.